### PR TITLE
[luci] Introduce FoldDepthwiseConv2D pass

### PR DESCRIPTION
**Pull request by ai-moiseev**
This commit introduces FoldDepthwiseConv2D pass for folding of DepthwiseConv2D nodes with constant inputs.

For #7505 
Requires mio-tflite: #7581

Signed-off-by: Alexander Moiseev <a.moiseev@samsung.com>

### CODE REVIEW

File: compiler/luci/pass/src/FoldDepthwiseConv2DPass.cpp [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                      node->stride()->h(), node->dilation()->h()))
+    return false; // Unsupported output parameters
+
+  if (!compute_output(&output_width, node->padding(), input_width, filter_width,
+                      node->stride()->h(), node->dilation()->h()))
+    return false; // Unsupported output parameters
+
+  auto const padding_height = compute_padding(node->stride()->h(), node->dilation()->h(),
+                                              input_height, filter_height, output_height);
+  auto const padding_width = compute_padding(node->stride()->h(), node->dilation()->h(),
```
Comment (jinevening) [inline compiler/luci/pass/src/FoldDepthwiseConv2DPass.cpp:163]: Maybe ?

```suggestion
  auto const padding_width = compute_padding(node->stride()->w(), node->dilation()->w(),
```

File: compiler/luci/pass/src/FoldDepthwiseConv2DPass.cpp [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+      WARN(l) << "Unsupported activation: " << uint32_t(node->fusedActivationFunction());
+      return false;
+    }
+  }
+
+  params->stride_height = node->stride()->h();
+  params->stride_width = node->stride()->w();
+  params->dilation_height_factor = node->dilation()->h();
+  params->dilation_width_factor = node->dilation()->w();
+  params->depth_multiplier = node->depthMultiplier();
```
Comment (jinevening) [inline compiler/luci/pass/src/FoldDepthwiseConv2DPass.cpp:96]: It would be better to add checks for multiplier.

```cpp
CHECK(num_filter_channels % num_input_channels == 0);
CHECK(depth_multiplier == num_filter_channels / num_input_channels);
```

File: compiler/luci/pass/CMakeLists.txt [hunk 3]
```diff
@@ -13,6 +13,7 @@ target_link_libraries(luci_pass PRIVATE luci_log)
 target_link_libraries(luci_pass PRIVATE luci_service)
 target_link_libraries(luci_pass PRIVATE luci_logex)
 target_link_libraries(luci_pass PRIVATE luci_profile)
+target_link_libraries(luci_pass PRIVATE mio_tflite_inc)
```
Comment (seanshpark) [inline compiler/luci/pass/CMakeLists.txt:16]: Could you please separate this change into another PR?
This is not related to introducing FoldDepthwiseConv2D Pass.

File: compiler/luci/pass/src/FoldDepthwiseConv2DPass.test.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  luci::CircleConst *getFoldedPattern() final
+  {
+    return loco::must_cast<luci::CircleConst *>(_output->from());
+  }
+
+protected:
+  luci::CircleDepthwiseConv2D *_dconv;
+  luci::CircleConst *_dconv_input;
+  luci::CircleConst *_dconv_filter;
+  luci::CircleConst *_dconv_bias;
```
Comment (seanshpark) [inline compiler/luci/pass/src/FoldDepthwiseConv2DPass.test.cpp:92]: I think our static analysis tool may require ` = nullptr` for every variables.

File: compiler/luci/requires.cmake [hunk 5]
```diff
@@ -5,6 +5,7 @@ require("locop")
 require("logo")
 require("logo-core")
 require("mio-circle")
+require("mio-tflite")
```
Comment (seanshpark) [inline compiler/luci/requires.cmake:8]: Please separate this to anoter PR too.

File: compiler/luci/pass/CMakeLists.txt (hunk 3 above)
Comment (ai-moiseev) [reply compiler/luci/pass/CMakeLists.txt:16]: Sure, I was planning to un-draft #7581 after mio-tflite PR merge.

File: compiler/luci/requires.cmake (hunk 5 above)
Comment (ai-moiseev) [reply compiler/luci/requires.cmake:8]: PTAL #7581

File: compiler/luci/pass/src/FoldDepthwiseConv2DPass.test.cpp (hunk 4 above)
Comment (ai-moiseev) [reply compiler/luci/pass/src/FoldDepthwiseConv2DPass.test.cpp]: With high chance it will, fixed, thank you :)

File: compiler/luci/pass/src/FoldDepthwiseConv2DPass.cpp (hunk 1 above)
Comment (ai-moiseev) [reply compiler/luci/pass/src/FoldDepthwiseConv2DPass.cpp]: Thank you! Fixed.

File: compiler/luci/pass/src/FoldDepthwiseConv2DPass.cpp (hunk 2 above)
Comment (ai-moiseev) [reply compiler/luci/pass/src/FoldDepthwiseConv2DPass.cpp:96]: Done, added these two checks.

**Review by seanshpark — DISMISSED** [review]
LGTM thank you!

**Review by jinevening — DISMISSED** [review]
LGTM

**Comment by seanshpark** [comment]
```
17:20:32 /opt/jenkins_slave/workspace/nnfw/master/pr-nncc-debug/compiler/luci/pass/src/FoldDepthwiseConv2DPass.cpp: In function ‘bool {anonymous}::fold_depthwise_conv_2d(luci::CircleDepthwiseConv2D*)’:
17:20:32 /opt/jenkins_slave/workspace/nnfw/master/pr-nncc-debug/compiler/luci/pass/src/FoldDepthwiseConv2DPass.cpp:151:31: error: comparison of integer expressions of different signedness: ‘int32_t’ {aka ‘int’} and ‘unsigned int’ [-Werror=sign-compare]
17:20:32    if (node->depthMultiplier() != filter_channels_out / input_depth)
17:20:32        ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17:20:32 cc1plus: all warnings being treated as errors
```

**Comment by seanshpark** [comment]
@hseok-oh , can you review too? or just skip?

**Comment by hseok-oh** [comment]
@seanshpark Please skip my review

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by jinevening — APPROVED** [review]
